server/src/main/java/org/candlepin/controller/Entitler.java

```diff
@@ -250,11 +254,44 @@ public void adjustEntitlementQuantity(Consumer consumer, Entitlement ent,
             devPool = poolManager.createPool(assembleDevPool(consumer, sku));
             data.setPossiblePools(Arrays.asList(devPool.getId()));
         }
-
         // Attempt to create entitlements:
         try {
             // the pools are only used to bind the guest
-            List<Entitlement> entitlements = poolManager.entitleByProducts(data);
+            // Here is where to get the pools that are provided by the host and attempt to entitleByProducts on those
+            // only, then we should check the compliance of the guest, if no good, entitleByProducts again, this time
+            // without the host pools.
+            // We should only attempt to use the host pools if there have not already been added any pools.
+            List<Entitlement> entitlements = new ArrayList<Entitlement>();
+            if (consumer.hasFact("virt.uuid") && "true".equalsIgnoreCase(consumer.getFact("virt.is_guest"))
+                && data.getPossiblePools().isEmpty()) {
+                // This is a guest that is trying to auto bind (either due to guest migration or possibly as a direct
+                // request. We should attempt first to auto bind using the pools provided by the current host.
+                Consumer host = consumerCurator.getHost(consumer.getFact("virt.uuid"), owner);
+                if (host != null) {
+                    List<String> hostProvidedPools = poolCurator.getPoolsProvidedByHost(host, owner);
+                    if (!hostProvidedPools.isEmpty()) {
+                        data.setPossiblePools(hostProvidedPools);
+                        entitlements.addAll(poolManager.entitleByProducts(data));
+                        // Check compliance to see if the host provided pools covered the guest entirely
+                        ComplianceStatus status = complianceRules.getStatus(consumer);
+                        if (!status.isCompliant()) {
+                            // We must not be compliant so we'll try to bind by products
+                            data.setPossiblePools(null);
+                            // Only auto attach for the non-compliant and partially compliant products
+                            Set<String> remainingProductsToCheck = status.getNonCompliantProducts();
+                            remainingProductsToCheck.addAll(status.getPartiallyCompliantProducts().keySet());
+                            data.setProductIds(remainingProductsToCheck.toArray(new String[remainingProductsToCheck.size()]));
+                            entitlements.addAll(poolManager.entitleByProducts(data));
+                        }
+                    } else {
```

Comment:
any chance we can reduce the size of this if else ladder using a boolean or perhaps turning the if else ladder around? looks like we are doing the same operation under a lot of conditional branching.